props/blaster.h

```diff
@@ -0,0 +1,192 @@
+  bool armed_ = false;
+
+  enum NextAction {
+    NEXT_ACTION_NOTHING,
+    NEXT_ACTION_ARM,
+    NEXT_ACTION_BLOW,
+  };
+
+  NextAction next_action_ = NEXT_ACTION_NOTHING;
+  uint32_t time_base_;
+  uint32_t next_event_time_;
+  
+  void SetNextAction(NextAction what, uint32_t when) {
+    time_base_ = millis();
+    next_event_time_ = when;
+    next_action_ = what;
+  }
+
+  void SetNextActionF(NextAction what, float when) {
+    SetNextAction(what, when * 1000);
+  }
+  
+  void PollNextAction() {
+    if (millis() - time_base_ > next_event_time_) {
+      switch (next_action_) {
+        case NEXT_ACTION_NOTHING:
+          break;
+        case NEXT_ACTION_ARM:
+          armed_ = true;
+          // TODO: Should we have separate ARMING and ARMED states?
+          break;
+        case NEXT_ACTION_BLOW:
+          Off(OFF_BLAST);
+          break;
+      }
+      next_action_ = NEXT_ACTION_NOTHING;
+    }
+  }
+  
+  void beginArm() {
+    SaberBase::SetLockup(SaberBase::LOCKUP_ARMED);
+    SaberBase::DoBeginLockup();
+#ifdef ENABLE_AUDIO    
+    float len = hybrid_font.GetCurrentEffectLength();
+#else    
+    float len = 1.6;
+#endif
+    SetNextActionF(NEXT_ACTION_ARM, len);
+  }
+
+  void selfDestruct() {
+    SaberBase::DoEndLockup();
+#ifdef ENABLE_AUDIO    
+    float len = hybrid_font.GetCurrentEffectLength();
+#else    
+    float len = 0.0;
+#endif
+    SaberBase::SetLockup(SaberBase::LOCKUP_NONE);
+    if (armed_) {
+      SetNextActionF(NEXT_ACTION_BLOW, len);
+    } else {
+      SetNextAction(NEXT_ACTION_NOTHING, 0);
+    }
+  }
+
+  void Loop() override {
+    PropBase::Loop();
+    PollNextAction();
+  }
+
+  // Make clash do nothing
+  void Clash(bool stab) override {}
+
+  // Make swings do nothing
+  void DoMotion(const Vec3& motion, bool clear) override { }
+
+  bool Event2(enum BUTTON button, EVENT event, uint32_t modifiers) override {
+    switch (EVENTID(button, event, modifiers)) {
+      case EVENTID(BUTTON_POWER, EVENT_LATCH_ON, MODE_OFF):
+      case EVENTID(BUTTON_AUX, EVENT_LATCH_ON, MODE_OFF):
+      case EVENTID(BUTTON_AUX2, EVENT_LATCH_ON, MODE_OFF):
+      case EVENTID(BUTTON_POWER, EVENT_CLICK_SHORT, MODE_OFF):
+        armed_ = false;
+        SetPower(true);
+        On();
+        return true;
+
+      case EVENTID(BUTTON_POWER, EVENT_LATCH_OFF, MODE_ON):
+      case EVENTID(BUTTON_POWER, EVENT_LATCH_OFF, MODE_OFF):
+      case EVENTID(BUTTON_POWER, EVENT_CLICK_LONG, MODE_ON):
+        SetPower(false);
+        Off();
+        return true;
+
+      case EVENTID(BUTTON_AUX, EVENT_DOUBLE_CLICK, MODE_OFF):
+        if (powered_) rotate_presets();
+        return true;
+
+      case EVENTID(BUTTON_POWER, EVENT_DOUBLE_CLICK, MODE_ON):
+        StartOrStopTrack();
+        return true;
+
+      case EVENTID(BUTTON_AUX, EVENT_PRESSED, MODE_ON):
+        Fire();
+        return true;
+
+      case EVENTID(BUTTON_AUX, EVENT_RELEASED, MODE_ON):
+        if (blaster_mode == MODE_AUTO) SaberBase::DoEndLockup();

```

Comment:
What happens if the mode changes between pressing and releasing AUX?

Reply:
Fixed. Would appreciat thoughts.. what I did was add a firing_ bool and don't allow mode changes during the firing sequence.

Reply:
In that case, maybe this if statement should check if (firing_) ?

Reply:
Well, we need to check the mode being auto. The firing_ set to true happens in the Fire() function and set to false on release.

What I wound up doing is NOT allowing a mode change if firing_